### ISSUE TITLE
Single-select form dropdown selected-value on many-to-many related models

### ISSUE THREAD

**Issue by robbykrlos** [laravel-enso/forms#91]
This is bug**.

### Prerequisites
* [x] Are you running the latest version?
* [x] Are you reporting to the correct repository?
* [x] Did you check the documentation?
* [x] Did you perform a cursory search?

### Description
Observer the following scenario: 
There are 2 models, related by many-to-many relation - so database has ``users``, ``cars``, ``car_user`` (relations table):
```php
class User
{
....
public function cars()
    {
        return $this->belongsToMany(Car::class);
    }
...
}
```
And this is used inside multiple forms as a multi-select dropdown where a user can be linked with multiple cars.

This is configured from :
``app/Forms/Templates/Models/user.json``
```json
{
                    "label": "Cars",
                    "name": "cars",
                    "value": [],
                    "meta": {
                        "type": "select",
                        "multiple": true,
                        "source": "models.cars.options"
                    }
                },
```

But, in some situations, some forms will require only single selection for this scenario. Should have been as easy as changing the 
`` "multiple": true`` to `` "multiple": false``.

But things are not so simple. Even though UI reacts as single-selection, loading existing (stored) values for the vue-select component is not working anymore.

Comparing the response from how the edit form is preparing data for one true single-select field which has one-to-many related models, we've observed that the data set on the ``edit`` response is different from what I get back on the many-to-many:

Ex: Single-select default menu inside roles form (one-to-many relation):
https://www.laravel-enso.com/system/roles/1/edit
![image](https://user-images.githubusercontent.com/16973022/98112388-cb5bd980-1eaa-11eb-8d5c-1a882527b6bd.png)
As seen here, value is integer "1".

Ex: Multi-select roles on user group form (many-to-many relation - table ``role_user_group``):
https://www.laravel-enso.com/administration/userGroups/1/edit
![image](https://user-images.githubusercontent.com/16973022/98112539-0cec8480-1eab-11eb-82c5-5e9d80187747.png)
As you can see here, value is an array (multi-select)

And now, here is what happens if the relation is many-to-many, and the ``multiple`` form meta flag is ``false``
![image](https://user-images.githubusercontent.com/16973022/98113669-abc5b080-1eac-11eb-8aaf-90614b12790a.png)
As you can see, the values set in this case are not expected id(s) single or multiple(array), but the whole json object is passed.

We've tracked this down to ``vendor/laravel-enso/forms/src/Services/Builder.php`` @90
```php
private function selectValue($value, $meta)
    {
        if ($meta->get('objects')) {
            return $value;
        }

        if ($meta->get('multiple')) {
            return $value instanceof Collection
                ? $value->pluck($meta->get('trackBy') ?? 'id')
                : $value;
        }

        return $value instanceof Model
            ? $value->{$meta->get('trackBy') ?? 'id'}
            : $value;
    }
```

specifically to this condition: 
```
 if ($meta->get('multiple')) {
```

My guess is that the assumption ``if not multiple`` will expect an integer value is wrong (expected from a one-to-many relation). The relation can still be a many-to-many, so and an array of objects will still be passed, which is not treated anymore with ``$value->pluck($meta->get('trackBy') ?? 'id')``

The trick we did, just to temporary fix this, was to create another meta element called ``many-to-many``
```php
"meta": {
                         ...
                        "type": "select",
                        "multiple": false,
                        "many-to-many": true,
                        ...
              }
```

Which we added to the condition:
```php
 if ($meta->get('multiple') || $meta->get('many-to-many')) {
            return $value instanceof Collection
                ? $value->pluck($meta->get('trackBy') ?? 'id')
                : $value;
}
```

We considered also using ```"multiple":"non-boolean-string-word-for-this-case"``` but this brings more complications.

We wanted to bring this into your attention, in case you consider it too as a *bug* or have better solutions that we overlooked.

Also, if you consider this a valid solution let me know and I'll PR it together with documentation - maybe come up with a better meta name for this situation - I'm picky with names, probably you too, and I'm still not satisfied with the "many-to-many" flag

**Comment by aocneanu** [comment]
The way forms & select are built now makes reading and writing in the database very clean, while the database respects the business logic.

I'm not 100% positive that I fully understood your use case, in fact I have a hard time  understanding a real example where you would have a mtm relation and want a single selection, or having a prop called `cars` in a form that will represent in fact a `car`.

But there will always be particular cases in particular projects and when things are not "natural" our approach is not to rely on conventions that can create confusions.

Coming back to your solution, it will break the FE select component which will receive an array as value but won't be multiple. At the first selection change the select will transform the value in a string / integer and some warnings will be thrown in the console. And you won't receive in BE an array, on submit, anyway. So probably you are doing something like `$user->cars()->sync([$request->get('cars')]` in your controller (notice the []).

What I would do instead, I would define the field named `car` in the form, because it actually holds the value of a car, and set in the MyForm class in the edit method `$this->form->value('car', $user->cars->first())` or something that makes sense in your context. And in the controller `$user->cars()->sync([$request->get('car')]`.

Makes things much readable imho.

**Comment by robbykrlos** [comment]
Hi,

Thank you for the fast and detailed explanations. I understand you point of view and where you are aiming with your alternative solution. We are considering it. I also agree with you that if things are done very clean in respect to db write/read, this exceptional case can be treated at another level.

On the other hand we still find our solution without issues. We are still investigating your suggested break points. 
I hope it's not one of those things that just work, because of too many coincidences...
Consider my following explanations as FYI (an "ahaa" moment), and not a pressure to accept this solution.

- In regards to the scenario. Yeah. can be wierd, but if 99% of the times users can have cars (ntn, multi-sleect) but one specific case where you would just one the FE to restrict to one selection (some users can have only one car... )

- in regards to the way data is received in ``app/Http/Controllers/Models/Users/Update.php``
```php 
public function __invoke(ValidateUserUpdateRequest $request, User $user)
    {
        dd($request->validated());
```
output:
```php
array:3 [
  "name" => "MyUser"
  "cars" => 4
]
```
so here the received selected "values" is not array but integer. Right. But...
This only interfere with validation which in this particular example has been corrected:
```php
public function rules()
    {
        return [
            'name'  => ['required', $this->nameUnique()],
            ...
            'cars' => 'integer', //from array
        ];
    }
```

- as for the sync - I was surprised we didn't had issues, and your suggestion made sense 
```php
public function updateWithCars(array $attributes)
    {
        tap($this)->update($attributes)
            ->cars()
            ->sync($attributes['cars']);
    }
```

but it seems sync is using : 
```php
parseIds($value) //where value is the given $ids
```

which in vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Concerns/InteractsWithPivotTable.php
```php
protected function parseIds($value)
    {
        if ($value instanceof Model) {
            return [$value->{$this->relatedKey}];
        }

        if ($value instanceof Collection) {
            return $value->pluck($this->relatedKey)->all();
        }

        if ($value instanceof BaseCollection) {
            return $value->toArray();
        }

        return (array) $value;
    }
```

is finally cast to array if none of the above cases are matching - so basically there is no change needed to adapt this new scenario.

- as for FE warnings - we got none. Console is clean on Form load, Form submit, Form reload.

- regarding: 
> Coming back to your solution, it will break the FE select component which will receive an array as value but won't be multiple

this is also something that makes sense, and now that you pointed out, yes, should have failed if a single-selection expects "integer" and not array, but for some reason - which I'm still looking into the @enso-ui/forms is still working. 

Thank you again for your assistance. Will close the ticket after your next reply.
